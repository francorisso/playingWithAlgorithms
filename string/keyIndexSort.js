export function keyIndexSort(arr, R) {
    const counter = new Array(R + 2).fill(0);
    arr.forEach(({ key }) => {
        counter[key + 1]++;
    });
    for (let i = 0; i < counter.length - 1; i++) {
        counter[i + 1] += counter[i];
    }

    const byKey = [];
    for (let i = 0; i < arr.length; i++) {
        byKey[counter[arr[i].key]++] = arr[i];
    }

    return byKey;
}
