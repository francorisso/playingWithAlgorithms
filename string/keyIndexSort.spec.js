import { keyIndexSort } from './keyIndexSort';

describe('keyIndexSort', () => {
    it('should sort by key', () => {
        expect(keyIndexSort([
            { name: 'A', key: 1 },
            { name: 'B', key: 2 },
            { name: 'C', key: 3 },
            { name: 'D', key: 1 },
            { name: 'F', key: 3 },
            { name: 'A2', key: 1 },
            { name: 'A3', key: 4 },
            { name: 'A4', key: 2 },
            { name: 'A5', key: 2 },
            { name: 'A6', key: 1 },
            { name: 'A7', key: 3 },
            { name: 'A8', key: 1 },
            { name: 'A3', key: 4 },
        ], 4)).toMatchSnapshot();
    });
});
