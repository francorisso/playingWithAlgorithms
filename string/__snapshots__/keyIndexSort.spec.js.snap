// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`keyIndexSort should sort by key 1`] = `
Array [
  Object {
    "key": 1,
    "name": "A",
  },
  Object {
    "key": 1,
    "name": "D",
  },
  Object {
    "key": 1,
    "name": "A2",
  },
  Object {
    "key": 1,
    "name": "A6",
  },
  Object {
    "key": 1,
    "name": "A8",
  },
  Object {
    "key": 2,
    "name": "B",
  },
  Object {
    "key": 2,
    "name": "A4",
  },
  Object {
    "key": 2,
    "name": "A5",
  },
  Object {
    "key": 3,
    "name": "C",
  },
  Object {
    "key": 3,
    "name": "F",
  },
  Object {
    "key": 3,
    "name": "A7",
  },
  Object {
    "key": 4,
    "name": "A3",
  },
  Object {
    "key": 4,
    "name": "A3",
  },
]
`;
